# Amazon Products 2023 (1.4M Products)

## Dataset
Due to the dataset exceeding 100 MiB which GitHub blocks files larger than 100 MiB please see the below link to the full dataset.

https://www.kaggle.com/datasets/asaniczka/amazon-products-dataset-2023-1-4m-products/data


Project Goal:
EDA with python

also create tableau dashboard.

Planning:

Conduct EDA for Amazon products
join the two datasets together.
identify patterns & trends 
explore  any correlations in the data
viz findings



# Project Objective
project ideas draft:
- which cateogires drive the most sales?
- Do customers ratings and reviews impact sales?
- Identify which niches are the easiest to make sales in?
- do customer ratings have much impact on sales?


Final objectives:
- <b>Analyse Sales Performance Across Categories:</b> Identify which product categories contribute the most to total sales.

- <b>Assess the Impact of Customer Ratings & reviews on Sales:</b> Investigate whether higher-rated products generate more sales compared to lower-rated ones.

- <b>Determine Profitable Niches:</b> Explore which niches have the highest conversion rates and consistent sales trends.

# Step 1: Import Libraries

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import csv

# Step 2: Load Datasets

In [3]:
# loading datasets
categories_df = pd.read_csv("C:\\Users\\Tim\\OneDrive\\Documents\\Projects\\Amazon Products 2023\\amazon_categories.csv")
products_df = pd.read_csv("C:\\Users\\Tim\\OneDrive\\Documents\Projects\\Amazon Products 2023\\amazon_products.csv")

In [4]:
# view first few rows of each dataset
categories_df.head()

id                     category_name
0   1          Beading & Jewelry Making
1   2                 Fabric Decorating
2   3       Knitting & Crochet Supplies
3   4              Printmaking Supplies
4   5  Scrapbooking & Stamping Supplies

In [5]:
products_df.head()

asin                                              title  \
0  B014TMV5YE  Sion Softside Expandable Roller Luggage, Black...   
1  B07GDLCQXV  Luggage Sets Expandable PC+ABS Durable Suitcas...   
2  B07XSCCZYG  Platinum Elite Softside Expandable Checked Lug...   
3  B08MVFKGJM  Freeform Hardside Expandable with Double Spinn...   
4  B01DJLKZBA  Winfield 2 Hardside Expandable Luggage with Sp...   

                                              imgUrl  \
0  https://m.media-amazon.com/images/I/815dLQKYIY...   
1  https://m.media-amazon.com/images/I/81bQlm7vf6...   
2  https://m.media-amazon.com/images/I/71EA35zvJB...   
3  https://m.media-amazon.com/images/I/91k6NYLQyI...   
4  https://m.media-amazon.com/images/I/61NJoaZcP9...   

                             productURL  stars  reviews   price  listPrice  \
0  https://www.amazon.com/dp/B014TMV5YE    4.5        0  139.99       0.00   
1  https://www.amazon.com/dp/B07GDLCQXV    4.5        0  169.99     209.99   
2  https://www.amazon.com/dp/B07XSCCZYG    4.6        0  365.49     429.99   
3  https://www.amazon.com/dp/B08MVFKGJM    4.6        0  291.59     354.37   
4  https://www.amazon.com/dp/B01DJLKZBA    4.5        0  174.99     309.99   

   category_id  isBestSeller  boughtInLastMonth  
0          104         False               2000  
1          104         False               1000  
2          104         False                300  
3          104         False                400  
4          104         False                400

# Step 3: Data Cleaning

In [6]:
# Check the number of rows and columns.
cat_rows = categories_df.shape[0] # Gives num of rows
cat_col = categories_df.shape[1] # gives num of cols
prod_row = products_df.shape[0] # Gives num of rows
prod_col = products_df.shape[1] # gives num of cols

print("categories_df Row Count: " + str(cat_rows))
print("cateogires_df column Count: " + str(cat_col))

print("products_df Row Count: " + str(prod_row))
print("products_df Column Count: " + str(prod_col))

categories_df Row Count: 248
cateogires_df column Count: 2
products_df Row Count: 1426337
products_df Column Count: 11


In [7]:
# Identify any missing values
print(categories_df.isnull().sum())
print(products_df.isnull().sum())

id               0
category_name    0
dtype: int64
asin                 0
title                1
imgUrl               0
productURL           0
stars                0
reviews              0
price                0
listPrice            0
category_id          0
isBestSeller         0
boughtInLastMonth    0
dtype: int64


In [8]:
# find the null value in products_df 
missing_title = products_df[products_df['title'].isnull()]
print(missing_title)

               asin title                                             imgUrl  \
1206102  B00758A9CE   NaN  https://m.media-amazon.com/images/I/71+50RU8o2...   

                                   productURL  stars  reviews  price  \
1206102  https://www.amazon.com/dp/B00758A9CE    5.0        0  33.08   

         listPrice  category_id  isBestSeller  boughtInLastMonth  
1206102        0.0           19         False                  0  


In [ ]:
# Use other information to try fill the missing title name.
prod_filter = products_df.loc[products_df['category_id'] == 19]
print(prod_filter)

               asin                                              title  \
1198836  B09DH98QL2  Chemical Guys HOL363 16-Piece Arsenal Builder ...   
1198837  B0C6DZZ4MN  Car Wash Kit -Car Wash Cleaning Kit Car Interi...   
1198838  B0CBZ12S15  Car Vacuum Cleaner High Power Cordless, 16000p...   
1198839  B08BRJHJF9  Microfiber Cleaning Cloth Grey - 12 packs 16"x...   
1198840  B06ZY896ZM  ThisWorx Car Vacuum Cleaner - Car Accessories ...   
...             ...                                                ...   
1207490  B093Q3RKB5  Car Cleaning Kit Buffing Pads Sponge Waxing an...   
1207491  B093CLN9YG  VViViD Essentials Car Cleaning, Waxing and Pol...   
1207492  B08RCV9DTM  Tires and Plastic Restoration - Professional C...   
1207493  B08HRMRJ95  Cleaning Washcloth Towels - Large 27" x 16" Mi...   
1207494  B08GG9W1WJ                Collinite 845 476s Double 520 Combo   

                                                    imgUrl  \
1198836  https://m.media-amazon.com/images/I/81xa

In [10]:
# Update column width to see full string of title column to make sure product title is being named 
# the same as what is shown in the productURL.
pd.set_option('display.max_colwidth', None)
with pd.option_context('display.max_colwidth', None):
    print(prod_filter)

               asin  \
1198836  B09DH98QL2   
1198837  B0C6DZZ4MN   
1198838  B0CBZ12S15   
1198839  B08BRJHJF9   
1198840  B06ZY896ZM   
...             ...   
1207490  B093Q3RKB5   
1207491  B093CLN9YG   
1207492  B08RCV9DTM   
1207493  B08HRMRJ95   
1207494  B08GG9W1WJ   

                                                                                                                                                                                                       title  \
1198836  Chemical Guys HOL363 16-Piece Arsenal Builder Ceramic Car Wash & Protection Kit with Big Mouth Max Release Foam Cannon, Bucket and (6) 16 oz Car Care Cleaning Chemicals (Works w/Pressure Washers)   
1198837                                          Car Wash Kit -Car Wash Cleaning Kit Car Interior Detailing Kit with Car Wash Foam Gun,Microfiber Sponge Car Care Kit,Tire Brush, Car Wash Kit with Foam Gun   
1198838          Car Vacuum Cleaner High Power Cordless, 16000pa Small Handheld Vacuum for Car with

In [ ]:
# Update Null value with product name from productURL.
# productURL is https://www.amazon.com/dp/B09DH98QL2?th=1

# Update the missing value with the given string from productURL
products_df['title'].fillna("Chemical Guys HOL363 16-Piece Arsenal Builder Ceramic Car Wash & Protection Kit with Big Mouth Max Release Foam Cannon, Bucket and (6) 16 oz Car Care Cleaning Chemicals (Works w/Pressure Washers)", inplace=True)

# Check if NULL value is replaced
print(products_df[products_df['title'].isnull()]) # should return an emptpy dataframe

# double checking
print(products_df.loc[products_df['asin'] == 'B00758A9CE']) # using unique asin from above

Empty DataFrame
Columns: [asin, title, imgUrl, productURL, stars, reviews, price, listPrice, category_id, isBestSeller, boughtInLastMonth]
Index: []
               asin  \
1206102  B00758A9CE   

                                                                                                                                                                                                       title  \
1206102  Chemical Guys HOL363 16-Piece Arsenal Builder Ceramic Car Wash & Protection Kit with Big Mouth Max Release Foam Cannon, Bucket and (6) 16 oz Car Care Cleaning Chemicals (Works w/Pressure Washers)   

                                                                 imgUrl  \
1206102  https://m.media-amazon.com/images/I/71+50RU8o2L._AC_UL320_.jpg   

                                   productURL  stars  reviews  price  \
1206102  https://www.amazon.com/dp/B00758A9CE    5.0        0  33.08   

         listPrice  category_id  isBestSeller  boughtInLastMonth  
1206102        0.0      

In [18]:
# Verify changes to check if anymore NULL values in products_df
print(products_df.isnull().sum())

asin                 0
title                0
imgUrl               0
productURL           0
stars                0
reviews              0
price                0
listPrice            0
category_id          0
isBestSeller         0
boughtInLastMonth    0
dtype: int64


#### Next Steps:
1. <b>Remove duplicates IDs:</b> Now that there are no NULL values in the dataset, the next step is to eliminate any duplicate entries in the 'id' column of categories_df to ensure data integrity.

2. <b>Validate Matching Keys:</b> Before merging the datasets, we need to verify that the 'id' values in categories_df align correctly with the 'category_id' values in products_df. This step ensures a smooth and accurate join.

3. <b>Merge the Datasets:</b> Once the key columns are validated, we will combine categories_df and products_df by joining on 'id' (from categories_df) and 'category_id' (from products_df). This will integrate category information with product details for further my final analysis and dataset used for my tableau dashboard.

In [13]:
# Check for duplicates in categories_df
categories_df.duplicated().sum()

np.int64(0)

In [14]:
products_df.duplicated().sum()

np.int64(0)

In [15]:
# Comparing id column from amazon_categories.csv to product_id in amazon_products.csv to ensure these categories are matching
cat_filter = categories_df.loc[categories_df['id'] == 19]
print(cat_filter)

prod_filter = products_df.loc[products_df['category_id'] == 19, ['title', 'category_id']]
print(prod_filter)

    id category_name
18  19      Car Care
                                                     title  category_id
1198836  Chemical Guys HOL363 16-Piece Arsenal Builder ...           19
1198837  Car Wash Kit -Car Wash Cleaning Kit Car Interi...           19
1198838  Car Vacuum Cleaner High Power Cordless, 16000p...           19
1198839  Microfiber Cleaning Cloth Grey - 12 packs 16"x...           19
1198840  ThisWorx Car Vacuum Cleaner - Car Accessories ...           19
...                                                    ...          ...
1207490  Car Cleaning Kit Buffing Pads Sponge Waxing an...           19
1207491  VViViD Essentials Car Cleaning, Waxing and Pol...           19
1207492  Tires and Plastic Restoration - Professional C...           19
1207493  Cleaning Washcloth Towels - Large 27" x 16" Mi...           19
1207494                Collinite 845 476s Double 520 Combo           19

[8659 rows x 2 columns]


In [16]:
uniquie_prod_id = products_df['category_id'].nunique()
print("Total unique category_id in products_df: " + str(uniquie_prod_id))

unique_cat_id = categories_df['id'].nunique()
print("Total unique id in categories_df: " + str(unique_cat_id))

Total unique category_id in products_df: 248
Total unique id in categories_df: 248


After comparing both categories_df & products_df datasets we can see that 'id' in categories_df matches the 'category_id' in the products_df after looking at specific cateogory names and product titles.

In [20]:
# Merge products_df with categories_df using a left join
merged_df = products_df.merge(categories_df, left_on = 'category_id', right_on = 'id', how = 'left')

# Checking if all categories have been successfully merged:
print(merged_df[['category_id','category_name']].drop_duplicates())

         category_id                           category_name
0                104                               Suitcases
924              110                          Men's Clothing
17994            245  Xbox 360 Games, Consoles & Accessories
21427            114                             Men's Shoes
38519            112                       Men's Accessories
...              ...                                     ...
1387414          204       Paint, Wall Treatments & Supplies
1392744          106                          Rain Umbrellas
1394518          105                      Travel Accessories
1414536          137     Stationery & Gift Wrapping Supplies
1414992           26           Car Electronics & Accessories

[248 rows x 2 columns]


the results shown above match the Total unique category_id in products_df: 248

In [ ]:
# Since id in categories_df is now redundant after merging, we can drop this column.
merged_df.drop(columns=['id'], inplace=True)

In [ ]:
# Preview merged_df to see if id column is dropped & if category_name column added.
merged_df.tail()

asin                                              title  \
1426332  B00R3LIKCO  American Flag Patriotic USA Classic 5 Panel Me...   
1426333  B098BQ7ZQ3  Men's Baseball Cap - H2O-DRI Line Up Curved Br...   
1426334  B07X1MVNT1  [4 Pack] Adjustable Eyeglasses and Sunglasses ...   
1426335  B08XLBG8V9                          Ax2002 Aviator Sunglasses   
1426336  B07GH67QC8  in Hoc Signo Vinces Knights Templar Masonic Em...   

                                                    imgUrl  \
1426332  https://m.media-amazon.com/images/I/71PDJFz6AA...   
1426333  https://m.media-amazon.com/images/I/812Tycexs4...   
1426334  https://m.media-amazon.com/images/I/61vvYW1S9J...   
1426335  https://m.media-amazon.com/images/I/51+yjD4F1x...   
1426336  https://m.media-amazon.com/images/I/91Kt2KQf0E...   

                                   productURL  stars  reviews  price  \
1426332  https://www.amazon.com/dp/B00R3LIKCO    4.2        0  14.95   
1426333  https://www.amazon.com/dp/B098BQ7ZQ3    4.4        0  33.99   
1426334  https://www.amazon.com/dp/B07X1MVNT1    3.6        0   8.54   
1426335  https://www.amazon.com/dp/B08XLBG8V9    4.5        0  54.36   
1426336  https://www.amazon.com/dp/B07GH67QC8    4.9        0  18.79   

         listPrice  category_id  isBestSeller  boughtInLastMonth  \
1426332       0.00          112         False                  0   
1426333       0.00          112         False                  0   
1426334       0.00          112         False                  0   
1426335      57.39          112         False                  0   
1426336       0.00          112         False                  0   

             category_name  
1426332  Men's Accessories  
1426333  Men's Accessories  
1426334  Men's Accessories  
1426335  Men's Accessories  
1426336  Men's Accessories

To help make my analysis easier and gather insights on whether: 
1. Sales Performance Across Categories
2. Impact of Customer Ratings & reviews on Sales

I will aggregate these fields and merge the aggegrated data & newly merged merged_df into a final dataset

In [ ]:
# Aggregate total sales per category

In [ ]:
# Aggregate ratings and reviews

In [ ]:

# check for duplicates based off id & product_id - done
# this includes checking for dup rows  in both datasets df.duplicated().sum() - done


# identify any missing values - done
# df.info() to see column type
# use describe to get more info
# If match we can join these two tables together making it easier to work with. - done 
# Then compare product_id and id to see if these cats match. this is last for sanity check to make sure. -done
# join 2 datasets
# drop columns which wont be using - done

